[#44945] Implementing `WorkPackage.at_timestamp`

fiedl opened the pull request:
In the context of the [baseline-comparison feature](https://community.openproject.org/projects/openproject/work_packages/26448), this pull request adds convenience methods to retrieve historic data of journable objects like work packages.

As the convenience methods introduced in this pull request are not used, this pull request does not change behaviour of the application, yet.

### Context

![Component_Overview-3](https://user-images.githubusercontent.com/1679688/217342096-88df5111-5825-452f-9bea-0a4aa639f3ea.png)

### Retrieve historic state of one journable

```ruby
WorkPackage.find(1).at_timestamp(1.year.ago)
```

This historic state of the work package is still a `WorkPackage` object, but it contains data from the `work_package_journals` table. It knows that it represents historic data and, therefore, cannot be modified.

```ruby
work_package = WorkPackage.find(1)
historic_work_package = work_package.at_timestamp(1.year.ago)

historic_work_package.id         # => 1  (same as work_package.id)
historic_work_package.historic?  # => true
historic_work_package.save       # raises `ActiveRecord::ReadOnlyRecord`
```

### Rollback to historic state of one journable

```ruby
WorkPackage.find(1).at_timestamp(1.year.ago).rollback!
```

This updates the work-package record with the attributes from the historic record, which does not remove journal entries, but, in fact, creates a new journal entry.

```ruby
WorkPackage.find(1).update! WorkPackage.find(1).at_timestamp(1.year.ago).attributes
```

### Retrieve historic state of multiple journables

```ruby
WorkPackage.at_timestamp(1.year.ago)
```

This returns an `ActiveRecord::Relation` that collects `WorkPackage` objects, which contain data from the `work_package_journals` table. These objects know that they represent historic data and, therefore, cannot be modified.

### Filter on historic data

```ruby
WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1)

# or, equivalently:
WorkPackage.where(assigned_to_id: 1).at_timestamp(1.year.ago)
```

This returns an `ActiveRecord::Relation` that represents all work packages that have been assigned to `User.find(1)` one year ago, regardless of whom they are assigned now.

### Plucking on historic data

```ruby
WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1).pluck :id
```

This returns the IDs of all `WorkPackage` records that have been assigned to `User.find(1)` one year ago.

These IDs are the IDs of the work-package records, not the work-package-journal records. Therefore, when fetching the work-package records from the database using these IDs, this will fetch the records in their current state, not the historic state.

```ruby
work_package_ids = WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1).pluck :id
work_packages = WorkPackage.where(id: work_package_ids)

work_packages.first.historic?  # => false
```

### Sub queries

```ruby
sub_query = WorkPackage.at_timestamp(1.year.ago).where(assigned_to_id: 1)
WorkPackage.where(id: sub_query)
```

This returns the IDs of all `WorkPackage` records that have been assigned to `User.find(1)` one year ago. As this is using a sub query, which is an `ActiveRecord::Relation`, this only hits the database once.

### Advanced queries

```ruby
WorkPackage.at_timestamp(1.year.ago).maximum(:estimated_hours)
```

This finds largest `estimated_hours` considering all work packages in their states one year ago.

```ruby
WorkPackage.at_timestamp(1.year.ago).group(:assigned_to_id).count  # => {1 => 15, ...}
```

This counts how many work packages have been assigned to each user one year ago.

```ruby
WorkPackage.at_timestamp(1.year.ago).group(:assigned_to_id).having('estimated_hours > 3').count  # => {1 => 12, ...}
```

This counts how many work packages have been assigned to each user one year ago, but only considers work packages with hour estimates over 3 hours.

### Cavests

#### Filtering on associated tables

When adding where conditions on associated tables, one might expect to also filter on historic data.

```ruby
# Does not work:
WorkPackage.at_timestamp(1.year.ago).joins(:time_entries).where(time_entries: {hours: 1.0})
```

Because `WorkPackage.at_timestamp` only searches the journal of the work packages, not the journal of the time entries, this does **not** return work packages that had time entries of 1 hour one year ago, but instead returns the state of the work packages of one year ago that do now have time entries of 1 hour.

#### Manual selects

Watch out when using the `id` column in manually specified parts of the query.

The wrapper introduced with `at_timestamp` internally modifies the active-record relation to query the work-package-journals table rather than the work-packages table. So, for example, when running

```ruby
WorkPackage.at_timestamp(1.day.ago).select("work_pacakges.id")
# raises `ActiveRecord::StatementInvalid: PG::UndefinedTable`
```

you would get an error because the `work_packges` table is not know to that query.

As a workaround, you may use the column `journables.id`, which is induced by the wrapper:

```ruby
WorkPackage.at_timestamp(1.day.ago).select("journables.id")
```

### Baseline comparison

Consider this pseudo API call to `GET /api/v3/work_packages` with the following parameters:

- `author_id: 1` specifies the author user of the work packges
- `offset: 2` specifies the page number of the paginated result
- `sortBy: [["status", "asc"]]` specifies the sort order to the result
- `timestamps: ["2021-09-04T13:21:14.452+00:00", "2022-09-04T13:21:14.452+00:00"]` specifies the timestamps at which historic data should be searched

The result should contain

- the work packages that match the filter `author_id` at **any** of the given `timestamps`
- the work-package data at **all** given `timestamps` in order for the user interface to show what has changed comparing the data at those `timestamps`.

Using the convenience methods introduced by this pull request, this is still a several-steps process:

```ruby
# given: params[:author_id], params[:offset], params[:sort_by], params[:timestamps]

# Apply filters
work_packages_relation = WorkPackage \
    .where(author_id: params[:author_id])

# Get IDs of work packages matching the filters at any of the given timestamps.
work_packages_for_all_timestamps = params[:timestamps].collect do |timestamp|
  work_packages_relation.at_timestamp(timestamp)
end.reduce(:or)

# Order the result
ordered_work_packages_relation = WorkPackage \
    .where(id: work_packages_for_all_timestamps) \
    .order(params[:sort_by])

# Paginate the result
paginated_work_packages_relation = \
    ordered_work_packages_relation \
    .paginate(page: params[:offset])
    
# Of all the matching work packages on that page, retrieve the historic attributes
# at all of the given timestamps.
paginated_work_packages_by_timestamp = {}
params[:timestamps].each do |timestamp|
  paginated_work_packages_by_timestamp[timestamp] = WorkPackage \
      .where(id: paginated_work_packages_relation) \
      .at_timestamp(timestamp)
end
```

### See also

- **Implementation Overview**: https://github.com/fiedl/openproject/issues/4
- **Implementation Tasks**: https://community.openproject.org/projects/openproject/work_packages/43483, https://community.openproject.org/projects/openproject/work_packages/44945
- **Epic**: https://community.openproject.org/projects/openproject/work_packages/26448
- https://github.com/fiedl/openproject/issues/3

fiedl committed:
implementing `Journal.at_timestamp`

in order to retrieve journal entries that correspond to a certain point in time.

If you have journals for Monday, Wednesday, and Friday, then `Journal.at_timestamp(Thursday)` will only include the Wednesday journal because that one is the most current on Thursday.

This is a step towards the baseline-queries feature:
- https://community.openproject.org/projects/openproject/work_packages/43483
- https://community.openproject.org/projects/openproject/work_packages/26448

fiedl committed:
implementing `Journable.at_timestamp`

in order to retrieve journable records with data from a certain point in time.

If you have work-package journals for Monday, Wednesday, and Friday, then `WorkPackage.at_timestamp(Thursday)` will retrieve the work-package records in their state of Wednesday, because this is the most current state on Thursday.

See: https://github.com/opf/openproject/pull/11243

fiedl committed:
Merge upstream branch 'dev' into fiedl:at_timestamp

https://github.com/opf/openproject/pull/11243

fiedl committed:
at_timestamp: adding reference to pull request

for documenting background information

https://github.com/opf/openproject/pull/11243

fiedl committed:
at_timestamp: adding missing time-entries attribute in specs

which has been introduced in https://github.com/opf/openproject/commit/4921e537c0bbc26749af288d5c0d0854f917e83d

fiedl committed:
at_timestamp: adding spec for sub queries

https://github.com/opf/openproject/pull/11243

fiedl committed:
at_timestamp: handle sql order clause

fiedl committed:
at_timestamp: handle sql where clause (as used by `Query#statement`)

fiedl committed:
at_timestamp: handle asking for work packages with specific ids

    WorkPackage.at_timestamp(...).where(id: ...)

fiedl committed:
to_timestamp: handle sub queries

https://github.com/fiedl/openproject/issues/3

https://github.com/opf/openproject/pull/11243

fiedl committed:
at_timestamp: refactoring with rubocop

fiedl committed:
at_timestamp: adding brief class documentation

fiedl committed:
Merge branch 'dev' into at_timestamp

fiedl commented:
@ulferts Ready for review.

This first pull request adds convenience methods, but does not call them. Therefore, this pull request does not change the behaviour of the application.

fiedl committed:
at_timestamp: handle where clauses with sql strings where the `work_packages.id` is referenced

which is used, for example, in 'follows'-relations queries.

fiedl committed:
at_timestamp: removing newline characters from hard-coded sql strings

fiedl committed:
at_timestamp: handle order clause with `work_package.id`

which is used in the manual-sorting feature.

fiedl committed:
at_timestamp: handle join clauses with `work_packages.id`

used by the manual-sorting feature.

fiedl committed:
at_timestamp: rubocop refactoring

fiedl committed:
at_timestamp: handle `Arel::Nodes::Grouping` (or clauses)

https://github.com/opf/openproject/pull/11678#issuecomment-1328011996

fiedl committed:
Merge branch 'dev' into https://github.com/opf/openproject/pull/11243

fiedl committed:
at_timestamp: handle `updated_at` and `created_at` in where clauses

as used in:

spec/features/work_packages/timeline/timeline_dates_spec.rb
spec/features/work_packages/table/queries/filter_spec.rb

https://github.com/opf/openproject/pull/11678#issuecomment-1328011996

Pull Request: https://github.com/opf/openproject/pull/11243

fiedl committed:
at_timestamp: handle case where "work_packages.foo IN array"

fiedl committed:
at_timestamp: moving some arel-related specs to historic_active_record_relation_spec

fiedl committed:
at_timestamp: handle case where "work_packages.id IN sub query"

https://github.com/opf/openproject/pull/11678#issuecomment-1324244907

Pull request: https://github.com/opf/openproject/pull/11243

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+    relation \
+        .joins("INNER JOIN \"journals\" ON \"journals\".\"data_type\" = '#{model.journal_class.name}'
+            AND \"journals\".\"data_id\" = \"#{model.journal_class.table_name}\".\"id\"".gsub("\n", "")) \
+        .merge(Journal.at_timestamp(timestamp))
```
There is no alias on this which might lead to conflicts.

fiedl replied:
related discussion: https://github.com/opf/openproject/pull/11243#issuecomment-1421456967

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+          predicate.left.name = "journable_id"
+          predicate.left.relation = Journal.arel_table
+        when "updated_at"
+          predicate.left.name = "created_at"
```
This should be kept as `updated_at`. We already talked about this so this is more a reminder.

fiedl replied:
fixed in https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

ulferts commented on app/models/journal/timestamps.rb:
```diff
@@ -0,0 +1,79 @@
+
+      timestamp = timestamp.in_time_zone if timestamp.kind_of? DateTime
+
+      where(id: where(created_at: (..timestamp)) \
```
Adding a limitation on the `journable_type` will reduce the number of journals to be processed. Unfortunately not by that much since 'WorkPackage' journals tend to be in the vast majority.

With it being added to the WHERE clause, it could be removed from both the ORDER BY as well as from the DISTINCT ON statement. However, tests on production data indicate that having both the added WHERE condition as well as keeping the column in the `distinct_on` leads to the best performance.

fiedl replied:
https://github.com/opf/openproject/pull/11243/commits/acf0d5580dbc3dfd0a5aa4b46f1e0f71401274dd

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+          #{model.journal_class.table_name}.*,
+          journals.journable_id as id,
+          journables.created_at as created_at,
+          journals.created_at as updated_at".gsub("\n", ""))
```
We already talked about  this but as a reminder: `journals.updated_at` should be used here instead.

ulferts replied:
Can we change `.gsub("\n", "")` to `.gsub(/\n\s+/, " ")`, or switch to a heredoc with `squish`? Admittedly it is nitpicking but it does read better in the console.

fiedl replied:
`updated_at`: fixed in fixed in https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+    relation
+  end
+
+  # Modify there where clauses such that e.g. the work-packages table is substituted
```
Typo: `there` -> `the`

ulferts commented on app/models/journable/timestamps.rb:
```diff
@@ -0,0 +1,103 @@
+          "updated_at" => journal.updated_at,
+          "timestamp" => timestamp,
+          "position" => (position if respond_to? :position)
+        }
```
The `lock_version` property is not included which leads to failures when doing:
` WorkPackage.first.at_timestamp(1.year.ago).lock_version`. The property could be mapped to the `version` column of the `journals` table.

fiedl replied:
Possibly related: https://github.com/opf/openproject/commit/e9c6bce775a3bbe53d7a2560a89232172114ee5b

With this change, the columns missing in the journal data are returned as `null`, which prevents errors, but does not provide data for these columns, yet.

@ulferts Could you explain what the `version` and the `lock_version` attributes are, and how they are related?

Do all journables have a `version` column? Then, this mapping could be added here; otherwise, I could add the modification to a subclass.

ulferts replied:
@fiedl , the `version` column is part of the `journals` table. So every journable will have it as all are using the `journals` table. The `lock_version` on the other hand is `WorkPackage` specific. It is part of [AR::OptimisticLocking](https://api.rubyonrails.org/classes/ActiveRecord/Locking/Optimistic.html).

So regarding your question, adding a mapping here can not be generalized as only WorkPackage has the `lock_version` column. 

In the current implementation, where `null` is returned from the DB, that value is unfortunately mapped to `0` in the ruby world as that is the default value for that column. 

I don't care too much about that column's information it is just that it might lead to surprising behaviour. So if we could design some mechanism that clearly communicates to not trust that value that would be good enough for me.

fiedl replied:
*Phone call 2023-02-02*:

- Retrieving the historic value for the `lock_version` column is not possible because this column is not journalized.
- We initialize the columns that are missing in the journalized table as `NULL` so that no database error is raised
- When trying to `pluck` or to retrieve such a value, we log a warning ➝ https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9, https://github.com/opf/openproject/pull/11243/commits/fe8ff8a1c12cf42bcebbda91d801808fb5c47036

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+    column_names.map! { |column_name| column_name == :id ? 'journals.journable_id' : column_name }
+    arel
+    super
+  end
```
There are more properties for which the behaviour might be surprising but those will fail hard:
* created_at
* updated_at
* lock_version
* position

From the top of my head, I don't think they are used within the codebase so it might be acceptable to let them simply fail hard until they are employed. On the other hand, it is a recurring pattern to have to map the properties so if some map could be established that is used in the various cases, further errors can be avoided. At least `created_at` and `updated_at` are non work package specific.

fiedl replied:
👍 resolved in https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+  def add_join_on_journables_table_with_created_at_column(relation)
+    relation \
+        .joins("INNER JOIN (SELECT id, created_at FROM \"#{model.table_name}\") AS journables
+            ON \"journables\".\"id\" = \"journals\".\"journable_id\"".gsub("\n", ""))
```
The `position` needs to also be included from here.

fiedl replied:
*Phone call 2023-02-02*:

- Retrieving the historic value for the `position` column is not possible because this column is not journalized.
- We initialize the columns that are missing in the journalized table as `NULL` so that no database error is raised
- When trying to `pluck` or to retrieve such a value, we log a warning ➝ https://github.com/opf/openproject/pull/11243/commits/41470cac7faebde554ebe98c7a1539adffef2bc9, https://github.com/opf/openproject/pull/11243/commits/fe8ff8a1c12cf42bcebbda91d801808fb5c47036

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+        array_of_joins.insert(0, journals_join)
+      end
+    end
+    arel
```
Instead of moving the join up afterwards, wouldn't it be easier to place the join at the first position in the `add_timestamp_condition` directly?

ulferts replied:
If not, I think it worthwhile to move the SQL string in `add_timestamp_condition` into its own method so that it can be called to compare the joins to.

fiedl replied:
Good catch, thanks! ➝ fixed in https://github.com/opf/openproject/pull/11243/commits/b838c6da17839c07a4ce3a82371a314119dc49ab

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+  #
+  def gsub_table_names_in_sql_string!(sql_string)
+    sql_string.gsub! /(?<!_)#{model.table_name}\.updated_at/, "journals.created_at"
+    sql_string.gsub! "\"#{model.table_name}\".\"updated_at\"", "\"journals\".\"created_at\""
```
Should both be switched to `journals.updated_at`

fiedl replied:
fixed in https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,290 @@
+    sql_string.gsub! /(?<!_)#{model.table_name}\.id/, "journals.journable_id"
+    sql_string.gsub! "\"#{model.table_name}\".\"id\"", "\"journals\".\"journable_id\""
+    sql_string.gsub! /(?<!_)#{model.table_name}\./, "#{model.journal_class.table_name}."
+    sql_string.gsub! "\"#{model.table_name}\".", "\"#{model.journal_class.table_name}\"."
```
Again, `lock_version` and `position` need to be mapped.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11243#discussion_r1099272555

ulferts reviewed (changes requested):
The smaller concerns you will find in the code. But the general code structure is good, so I don't expect them to cause major problems.

My one big concern though is performance. On one of our larger customer databases there are about 1 million work packages and over 5 million journals of which the vast majority are associated to a work package. A query for historic data will always have to sort the journals (e.g.

```
...
WHERE "journals"."id" IN (
SELECT distinct on (journals.journable_type, journable_id) id FROM "journals" 
 WHERE "journals"."created_at" <= '2022-01-06 07:57:14.153220' AND journals.journable_type = 'WorkPackage'
 ORDER BY journals.journable_type ASC, "journals"."journable_id" ASC, "journals"."created_at" DESC)
 ```
 
 ) to find the most recent journal older than the time queried for. And an increase in journals will almost certainly increase the number of journals to sort assuming that most users will be less interested in comparing to a very old state. 

I already tried a couple of things: 

* Changing to an `EXISTS` which might perform better if the work package journals can be filtered by different conditions (e.g. `assigned_to_id = 5`) or to an inner join.
* Different ways of finding the most recent journal (e.g. "group by" and self left join)
* Denormalizing data by duplicating columns from `journals` and `work_packages` in the `work_package_journals` table.
* Adding indices, to help with the ordering.
* Adding a partial index only for `journable_type = 'WorkPackages' 
* Having a materialized view to precalculate the historic data structure containing `journals`, `work_packages` and `work_package_journals`

In the end, all my attempts performed poorly in the worst case which is not filtering on the `work_package_journals` table at all. Unfortunately, that worst case is not unlikely to take place. And in that case, a historic query tends to perform an order of magnitude worse than a similar non historic one.

What I would propose to do is try to identify the most likely timestamps for which users would like to get the baseline (e.g. 1 week ago, 1 month ago, ...). For those. we could create materialized views on a regular basis and they would just contain the journals necessary to know within that timeframe which journal is the most recent one. That would not mean that there would be only one journal per work package but there would be way fewer than the complete set of journals and could be sorted. A query could then switch to the appropriate view based on the time frame queried for.

A benefit of that approach is that the code in this PR could be simplified but we would have to pay for this by having to maintain the materialized views.

Maybe you have more ideas @fiedl and if so, I would be very interested in talking about them.

fiedl commented:
## `updated_at` column

*Call 2022-12-21*

The `work_package_journals` table does not have an `updated_at` column as the `work_packages` table. When asking for the `updated_at` attribute, we need to map this to the `journals` table.

Originally, I have mapped the `updated_at` attribute to `journals.created_at`, assuming that each journal entry corresponds to a specific change. However, the journal entries can be updated when several changes are made in short succession in order to keep the activity log cleaner. 

- [x] Because journal entries are updated in this scenario, the last time of update is actually `journals.updated_at` rather than `journals.created_at`. ➝ https://github.com/opf/openproject/pull/11243/commits/d1962f14d441e8fb54f83c5bc4244c91badb0b5a

fiedl committed:
at_timestamp: correct mapping of the `updated_at` column

The `work_package_journals` table does not have an `updated_at` column as the `work_packages` table. When asking for the `updated_at` attribute, we need to map this to the `journals` table.

Originally, I have mapped the `updated_at` attribute to `journals.created_at`, assuming that each journal entry corresponds to a specific change. However, the journal entries can be updated when several changes are made in short succession in order to keep the activity log cleaner. Because journal entries are updated in this scenario, the last time of update is actually `journals.updated_at` rather than `journals.created_at`.

https://github.com/opf/openproject/pull/11243#issuecomment-1386919337

fiedl committed:
Merge branch 'dev' into https://github.com/opf/openproject/pull/11243

fiedl committed:
Journable::Timestamps: columns missing in the journal-data table are interpreted as null

https://github.com/opf/openproject/pull/11243

cherry-picked from https://github.com/opf/openproject/commit/e9c6bce775a3bbe53d7a2560a89232172114ee5b in the context of https://github.com/opf/openproject/pull/11783

fiedl committed:
at_timestamp: rubocop refactoring

https://github.com/opf/openproject/pull/11243

fiedl committed:
HistoricActiveRecordRelation#pluck: map existing columns and warn if column is not journalized

as suggested in the review by Jens Ulferts

https://github.com/opf/openproject/pull/11243/files#r1064582100

fiedl committed:
HistoricActiveRecordRelation#pluck: rubocop refactoring

https://github.com/opf/openproject/pull/11243/files#r1064582100

fiedl committed:
HistoricActiveRecordRelation: improve documentation of `substitute_database_table_in_predicate`

fiedl commented:
## Table alias for "journals" table

As suggested in https://github.com/opf/openproject/pull/11243#discussion_r1063195259, using a local table alias for the "journals" table would prevent accidental collisions when constructing a query like this:

```ruby
WorkPackage.at_timestamp(timestamp).includes(:journals).where(journals: {...})
```

However, introducing the alias is not trivial, I think:

```ruby
class Journable::HistoricActiveRecordRelation < ActiveRecord::Relation
  JOURNALS_TABLE_ALIAS = "journals2".freeze

  def add_timestamp_condition(relation)
    journal_relation = Journal.at_timestamp(timestamp)

    # TODO: journal_relation = journal_relation.with_table_alias(JOURNALS_TABLE_ALIAS)

    relation \
        .joins("INNER JOIN \"journals\" AS \"#{JOURNALS_TABLE_ALIAS}\"
            ON \"#{JOURNALS_TABLE_ALIAS}\".\"data_type\" = '#{model.journal_class.name}'
            AND \"#{JOURNALS_TABLE_ALIAS}\".\"data_id\" = \"#{model.journal_class.table_name}\".\"id\"".gsub("\n", "")) \
        .merge(journal_relation)
  end
end
```

Introducing the table alias in the sql literal above is no problem. But introducing the alias into the arel parts of the query has unintended side effects.

For example, when trying

```ruby
journal_relation.arel.source.left.table_alias = JOURNALS_TABLE_ALIAS
```

then the table alias is set not only for this relation, but on a deeper object within, i.e. `Journal.arel_table` itself. The above is equivalent to setting directly:

```ruby
Journal.arel_table.table_alias = JOURNALS_TABLE_ALIAS
```

This, unfortunately, has side effects on sql queries by `ActsAsJournalized` as well, e.g. `./lib_static/plugins/acts_as_journalized/lib/acts/journalized/reversion.rb:79`, which then would fail unless properly patched.

In conclusion:

- [ ] Either we find a way to set the table alias only for a single active-record relation,
- [x] or we postpone the attempt until we really need the table alias.

fiedl committed:
historic journables: raise warning when accessing non-journalized columns

https://github.com/opf/openproject/pull/11243/files#r1064584887

fiedl committed:
historic journables: prevent accessing column information on a non-migrated database

which does happen in ci: https://github.com/opf/openproject/actions/runs/4118578222/jobs/7111270504

fiedl committed:
HistoricActiveRecordRelation: prepend journals join rather than moving it up later

caught by Jens Ulferts: https://github.com/opf/openproject/pull/11243#discussion_r1099359693

fiedl committed:
HistoricActiveRecordRelation: refactoring multi-line sql string

fiedl committed:
Merge branch 'dev' into https://github.com/opf/openproject/pull/11243

fiedl committed:
HistoricActiveRecordRelation: include journable_type condition in subquery

in order to improve query performance

caught by Jens Ulferts: https://github.com/opf/openproject/pull/11243#discussion_r1063247444

https://github.com/opf/openproject/pull/11243

fiedl committed:
Merge branch 'dev' into https://github.com/opf/openproject/pull/11243

in order to get https://github.com/opf/openproject/commit/69d95650655195dd777810d9951864b567a22248 to fix a failing spec

fiedl committed:
Merge dev into https://github.com/opf/openproject/pull/11243

fiedl committed:
Merge branch 'dev' https://github.com/opf/openproject/pull/11243

ulferts reviewed (approved):
@fiedl I am quite happy about this by now. Looks good and I'll merge it.

There is one thing that is not problematic right away but might lead to surprising behaviour later on and that lies in a call like 

```
puts WorkPackage.at_timestamp(5.days.ago).select(:id).to_sql

=> SELECT "work_package_journals"."id" FROM "work_package_journals" INNER JOIN "journals" ON "journals"."data_type" = 'Journal::WorkPackageJournal' AND "journals"."data_id" = "work_package_journals"."id" INNER JOIN (SELECT id, created_at FROM "work_packages") AS journables ON "journables"."id" = "journals"."journable_id" WHERE "journals"."journable_type" = 'WorkPackage' AND "journals"."id" IN (SELECT distinct on (journable_type, journable_id) id FROM "journals" WHERE "journals"."journable_type" = 'WorkPackage' AND "journals"."created_at" <= '2023-02-10 15:27:31.301822' ORDER BY "journals"."journable_type" ASC, "journals"."journable_id" ASC, "journals"."created_at" DESC)

```

In such a statement, the projection will be `work_package_journals.id`. If such a statement were to be used in a subselect, the wrong IDs would be referenced.

Compare that to the way `pluck` is handled explicitly. 

```
SELECT "journals"."journable_id" FROM ...
```

Ideally, the same mapping should be applied to `select` as it is for `pluck`. 

Nevertheless, this is not an immediate problem which is why I am merging this PR now.
